### PR TITLE
Compare exchange

### PR DESCRIPTION
**Pull request by EnderGnar**

### CODE REVIEW

File: spec/lang/intrinsic.md [hunk 1]
```diff
@@ -300,3 +300,58 @@ impl<M: Memory> Machine<M> {
...
+        let size = ret_ty.size::<M>();
+        // ISSUE: Can't be tested because`CompareExchange` only works with integeres and all integers are power of two.
+        if !size.bytes().is_power_of_two() {
+            throw_ub!("invalid return type for `Intrinsic::CompareExchange`, size not power of two");
+        }
+        if size > M::MAX_ATOMIC_SIZE {
+            throw_ub!("invalid return type for `Intrinsic::CompareExchange`, size to big");
+        }
+        
+        let pty = PlaceType{ty: ret_ty, align: Align::max_for_offset(size).unwrap()};
```
Comment (RalfJung) [inline spec/lang/intrinsic.md:343]: - please add spaces
- please use `Align::from_bytes`

File: spec/lang/intrinsic.md [hunk 2]
```diff
@@ -300,3 +300,58 @@ impl<M: Memory> Machine<M> {
...
+        if curr_ty != ret_ty {
+            throw_ub!("invalid second argument to `Intrinsic::CompareExchange`, not same type");
+        }
+        if next_ty != ret_ty {
+            throw_ub!("invalid third argument to `Intrinsic::CompareExchange`, not same type");
+        }
+
+        let size = ret_ty.size::<M>();
+        // ISSUE: Can't be tested because`CompareExchange` only works with integeres and all integers are power of two.
+        if !size.bytes().is_power_of_two() {
```
Comment (RalfJung) [inline spec/lang/intrinsic.md:336]: can be an assert

File: tooling/minitest/src/ub/compare_exchange.rs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+#[test]
+fn compare_exchange_success() {
+    let locals = [ <u32>::get_ptype(); 2 ];
+
+    let ptr_ty = raw_ptr_ty( <u32>::get_layout() );
+
+    let addr0 = addr_of(local(0), ptr_ty);
+
+    // Can overwrite.
```
Comment (RalfJung) [inline tooling/minitest/src/ub/compare_exchange.rs:11]:
```suggestion
    // Success case: check that we do perform a store.
```

File: spec/lang/intrinsic.md [hunk 4]
```diff
@@ -300,3 +300,55 @@ impl<M: Memory> Machine<M> {
...
+        }
+        if curr_ty != ret_ty {
+            throw_ub!("invalid second argument to `Intrinsic::CompareExchange`, not same type");
+        }
+        if next_ty != ret_ty {
+            throw_ub!("invalid third argument to `Intrinsic::CompareExchange`, not same type");
+        }
+
+        let size = ret_ty.size::<M>();
+        assert!(size.bytes().is_power_of_two());
```
Comment (RalfJung) [inline spec/lang/intrinsic.md:335]:
```suggestion
        // All integer sizes are powers of two.
        assert!(size.bytes().is_power_of_two());
```

File: tooling/minitest/src/ub/compare_exchange.rs [hunk 5]
```diff
@@ -22,9 +22,9 @@ fn compare_exchange_success() {
         print(load(local(1)), 3)
     );
 
-    // Doesn't if current doesn't match.
+    // Failure case: check that we do not perform a store
```
Comment (RalfJung) [inline tooling/minitest/src/ub/compare_exchange.rs:25]: The test doesn't actually print `local(0)` again, does it? So it doesn't check whether a store happened or not?

File: tooling/minitest/src/ub/compare_exchange.rs [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+    // Success case: check that we do perform a store.
+    let b0 = block!(
+        storage_live(0),
+        storage_live(1),
+        assign(local(0), const_int::<u32>(0)),
+        compare_exchange(local(1), addr0, const_int::<u32>(0), const_int::<u32>(1), 1),
+    );
+    let b1 = block!(
+        print(load(local(0)), 2)
```
Comment (RalfJung) [inline tooling/minitest/src/ub/compare_exchange.rs:19]:
```suggestion
        // print value of CASed location
        print(load(local(0)), 2)
```

File: tooling/minitest/src/ub/compare_exchange.rs [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    );
+    let b2 = block!(
+        print(load(local(1)), 3)
```
Comment (RalfJung) [inline tooling/minitest/src/ub/compare_exchange.rs:22]:
```suggestion
        // print CAS return value
        print(load(local(1)), 3)
```

File: tooling/minitest/src/ub/compare_exchange.rs [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    );
+
+    // Failure case: check that we do not perform a store
+    let b3 = block!(
+        compare_exchange(local(1), addr0, const_int::<u32>(3), const_int::<u32>(42), 4)
+    );
+    let b4 = block!(
+        print(load(local(1)), 5)
```
Comment (RalfJung) [inline tooling/minitest/src/ub/compare_exchange.rs:30]:
```suggestion
        // print CAS return value
        print(load(local(1)), 5)
```

File: tooling/minitest/src/ub/compare_exchange.rs (hunk 5 above)
Comment (EnderGnar) [reply tooling/minitest/src/ub/compare_exchange.rs:25]: I removed the wrong print.. Should I once again check both or is it enough to check that `local(0)` did not change?

Comment (RalfJung) [reply tooling/minitest/src/ub/compare_exchange.rs:27]: Seems reasonable to check both, why not.

File: spec/lang/intrinsic.md [hunk 9]
```diff
@@ -300,3 +300,56 @@ impl<M: Memory> Machine<M> {
     }
 }
 ```
+
+Some more complex atomic memory operations are needed as well.
+
+```rust
+impl<M: Memory> Machine<M> {
```
Comment (RalfJung) [inline spec/lang/intrinsic.md:307]:
```suggestion

```
no reason to have these in separate blocks, they all together form the atomic memory operations

**Pull request merged**